### ISSUE TITLE
T-style intersections

### ISSUE THREAD

**Issue by impworks** [OpenKieler/elkjs#54]
Is it possible to force ELK to create T-style intersections?

I'm trying to generate a layout for a family tree using elk.js.
This is the expected layout:
![image](https://user-images.githubusercontent.com/604496/48357065-be83c100-e6a8-11e8-9f63-a4a1b940cce3.png)

Actual layout is a bit off:
![image](https://user-images.githubusercontent.com/604496/48357088-d22f2780-e6a8-11e8-8313-ac1a2d443411.png)

Here's the JSON I am using:

```
{
    "id": "root",
    "layoutOptions": {
        "elk.algorithm": "layered",
        "elk.direction": "DOWN",
        "elk.edgeRouting": "ORTHOGONAL",
        "elk.layered.mergeEdges": true
    },
    "children": [
        {
            "id": "P1",
            "width": 100,
            "height": 50,
            "layoutOptions": { "elk.portConstraints": "FIXED_SIDE" },
            "ports": [
                { "id": "P1:e", "layoutOptions": { "elk.port.side": "EAST" } },
                { "id": "P1:w", "layoutOptions": { "elk.port.side": "WEST" } }
            ]
        },
        {
            "id": "P2",
            "width": 100,
            "height": 50,
            "layoutOptions": { "elk.portConstraints": "FIXED_SIDE" },
            "ports": [
                { "id": "P2:e", "layoutOptions": { "elk.port.side": "EAST" } }, 
                { "id": "P2:w", "layoutOptions": { "elk.port.side": "WEST" } }
            ]
        },
        { 
            "id": "P3",
            "width": 100,
            "height": 50,
            "layoutOptions": { "elk.portConstraints": "FIXED_SIDE" },
            "ports": [
                { "id": "P3:n", "layoutOptions": { "elk.port.side": "NORTH" } }
            ]
        },
        {
            "id": "P4",
            "width": 100,
            "height": 50,
            "layoutOptions": { "elk.portConstraints": "FIXED_SIDE" },
            "ports": [
                { "id": "P4:n", "layoutOptions": { "elk.port.side": "NORTH" } }
            ]
        },
        {
            "id": "P5",
            "width": 100,
            "height": 50,
            "layoutOptions": { "elk.portConstraints": "FIXED_SIDE" },
            "ports": [
                { "id": "P5:n", "layoutOptions": { "elk.port.side": "NORTH" } }
            ]
        },
        {
            "id": "R1",
            "width": 1,
            "height": 1,
            "layoutOptions": { "elk.portConstraints": "FIXED_SIDE" }, 
            "ports": [
                { "id": "R1:n", "layoutOptions": { "elk.port.side": "NORTH" } },
                { "id": "R1:s", "layoutOptions": { "elk.port.side": "SOUTH" } }
            ]
        }
    ],
    "edges": [
        { "id": "E1", "sources": [ "P1:e" ], "targets": [ "R1:n" ] },
        { "id": "E2", "sources": [ "P2:w" ], "targets": [ "R1:n" ] },
        { "id": "E3", "sources": [ "R1:s" ], "targets": [ "P3:n" ] },
        { "id": "E4", "sources": [ "R1:s" ], "targets": [ "P4:n" ] },
        { "id": "E5", "sources": [ "R1:s" ], "targets": [ "P5:n" ] }
    ]
}
```

**Comment by le-cds** [comment]
No, we can't, due to eclipse/elk#328. I'll link back to this issue for another set of examples that demonstrate parts of that problem.

**Comment by impworks** [comment]
Maybe there's a way to force the fake connector node to be positioned in the same row as the parent nodes, like the `[constraint=false]` directive in [DOT](https://www.graphviz.org/doc/info/attrs.html#d:constraint)? Something like this:

![image](https://user-images.githubusercontent.com/604496/48401529-7614e380-e73a-11e8-9a51-fa4fd005b806.png)

**Comment by le-cds** [comment]
There are two fake connector nodes there, one each for `P1` and `P2`. The solution will be to allow them to be merged. :)

**Comment by EmilStenstrom** [comment]
I’ve been looking all over for a layout engine that supports family trees, and it’s really thorny when you get down to practical trees. Would love to see a fix for these T-shaped connectors.

**Comment by impworks** [comment]
@EmilStenstrom just in case you're wondering, I solved the problem for my case by using a layout like this:

![image](https://user-images.githubusercontent.com/604496/87918697-791fb580-ca7f-11ea-8821-75515a3f7ee8.png)

Where the red square is an invisible pseudo-node, which has a size of 1x1 pixels and two ports: north (for parents) and south (for descendants). Although it's not exactly what I wanted, it seems to be working pretty well.

**Comment by EmilStenstrom** [comment]
@impworks looks great! Thank you!

**Comment by EmilStenstrom** [comment]
@impworks What graphing library do you use? Is it web based? I can't find any examples of using elkjs with d3 for instance (but I see that Klayjs has a d3 adapter).

**Comment by impworks** [comment]
@EmilStenstrom I didn't need any library. Since ElkJS does all the calculations, all that's left is to display stuff on the page at given points. For the sample images above I used Canvas API, and for the real project I use a combination of SVG (for lines) and a lot of absolutely-positioned `div` tags for blocks. Here's a working example (it's in Russian though, but you get the idea):

http://bonsai.kirillorlov.pro/p/%D0%A1%D0%B5%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0_%D0%98%D1%80%D0%B8%D0%BD%D0%B0_%D0%90%D0%BB%D0%B5%D0%BA%D1%81%D0%B5%D0%B5%D0%B2%D0%BD%D0%B0/tree

**Comment by EmilStenstrom** [comment]
@impworks Thanks! It seems need to write a graphing library myself :) was hoping that someone else had written an adapter that worked well :)

**Comment by EmilStenstrom** [comment]
> I solved the problem for my case by using a layout like this:
> ![image](https://user-images.githubusercontent.com/604496/87918697-791fb580-ca7f-11ea-8821-75515a3f7ee8.png)
> 
> Where the red square is an invisible pseudo-node, which has a size of 1x1 pixels and two ports: north (for parents) and south (for descendants). Although it's not exactly what I wanted, it seems to be working pretty well.

@impworks how did you get it centered? I used your example code from above, and changed the ports, and what I get is this: 
<img width="367" alt="image" src="https://user-images.githubusercontent.com/224130/93721374-0ff81380-fb90-11ea-8a73-a2257e904f30.png">

**Comment by uruuru** [comment]
Try `nodePlacement.strategy: NETWORK_SIMPLEX`.

**Comment by EmilStenstrom** [comment]
That just seems to put it to the right instead:

<img width="353" alt="image" src="https://user-images.githubusercontent.com/224130/93822039-de4e7d80-fc5f-11ea-9c13-a1367a85e89c.png">

Here's an [interactive version](http://rtsys.informatik.uni-kiel.de/elklive/json.html?compressedContent=N4KABBYEQJYCZQFzQE4Hs0BcoBpySgBsBDATzQFdMB5AB0xjQDsBnJMUSL6YwgczQoYmABYBbdkTIBTFNIR5uBaYQDWAOjgw5AYwbNJAEWoB1AHK58SqEzRxpABRI7pY6U0zqWmFMUzS+UkkzAFEAFRNqACUAaQB9AGUASQBZBwAZEIANKCswAF9FAh0RGEI4OSZ2AG08rk4lblgEZCgHAEZLRqaAd3hRdnaABiGi7ogoEWkYPhFsZABWUbrGqFpBTDZkWvHG4GbJDsQqnGgScio6fVZ2fZZ4aWDosIAJKHzCld39+EP2xDYpykFxo9EYN2QdwekgS1AAqq93vkvtwALpjcZSABGKi2YGqDV2XCg-gAHvNoB1ckSCqiUZ9uoTMb9Wg4AExdIlQPpwAbIYbLGmTaazClLDHdNYbPE7GmQH4tSls464M5kSig654qH2J5RREfCVEhWHZWAtUgq7g7VQe661qwhFvD4orjo10EEg4wgyplCskUtocj0QfJ07oMvYh6AsykAZk5u25-REgxGRusUxmc3Y4ujUpQmxq0fqB1ZcZVQPOGqtzBtdserTMz2dyJp7qFXtxNT9XIDhwT0bD9IzHHzsbaABZE5ieXywALR8Ss6Lc4KuetCzKS-Ky5TJ5WLTWwXXbrboU2W0jox2uV2fT2dxN+6zp0Pw41I0pe9YJw4FjOkpzqm-LpvmK45os65JpuRbbE+HB7m0CyHsCx5ameDZ6gabZEreSb3r6CEktI5KHAB74jiiP5NBOUSdEuBDAYMjETBBFLtKx0Cwducq7nR-wnEelwnhCiFYZe+qtlxpYCQCqpoSJGGQue9rQI6OE3vSeT4dA8h8NIvF7EhIQMdALCUCgLgypS-xsGA6LQJgxAoAZcH4tA9Eqg5BSjiarQhByQIWRQVmGTUSryQ5QLOa50judUnmCVAPlftw-nQCECbBZZ1kRVAXn2Y5JIuW5NltBWVSpX5JnTjloV5dsSVRcVsVlflDgHlVqK+dRJkAfVYXlYVKWtaV8Xlf+3k9bhEB0siQA)

**Comment by EmilStenstrom** [comment]
Oh! Seems SIMPLE and LINEAR_SEGMENTS achieved what I wanted! Thanks @uruuru!
